app/controllers/admin/deals_controller.rb

```diff
@@ -0,0 +1,70 @@
+class Admin::DealsController < ApplicationController
+  before_action :check_admin
+  before_action :set_deal, only: %i[show edit update destroy check_publishablity]
+  
+  def index
+    @deals = Deal.all
+  end
+
+  def new
+    @deal = Deal.new
+  end
+
+  def show
+  end
+
+  def edit
+  end
+
+  def create
+    @deal = Deal.new(deal_params)
+    respond_to do |format|
+      if @deal.save
+        format.html { redirect_to admin_deal_path(@deal), notice: "Deal was successfully created." }
+      else
+        format.html { render :new, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def update
+    respond_to do |format| 
+      if @deal.update(deal_params)
+        format.html { redirect_to admin_deal_url(@deal), notice: "Deal was successfully updated." }
+      else
+        format.html { render :new, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  def destroy
+    @deal.destroy
```

Comment:
Yes it can fail when an error is raised from db side, For example if other admin has locked it in db while updating it in that case it can't be accessed for deletion and destroy will return false.